src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,203 @@
+        } else {
+            temp.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            addBefore(value, searchNode(index));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public void addAll(List<T> list) {
+        Object[] items = list.toArray();
+        int numNew = items.length;
+        if (numNew == 0) {
+            return;
+        }
+        Node<T> beforeCurrent;
+        beforeCurrent = tail;
+        for (Object item : items) {
+            Node<T> newNode = new Node<>(beforeCurrent, (T) item, null);
+            if (beforeCurrent == null) {
+                head = newNode;
+            } else {
+                beforeCurrent.next = newNode;
+            }
+            beforeCurrent = newNode;
+        }
+        tail = beforeCurrent;
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return searchNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> current = searchNode(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(searchNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = head; current != null; current = current.next) {
+            if (object == null && current.item == null) {
+                unlink(current);
+                return true;
+            } else if (object != null && object.equals(current.item)) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+
+        public T getItem() {
+            return item;
+        }
+
+        public void setItem(T item) {
+            this.item = item;
+        }
+
+        public Node<T> getNext() {
+            return next;
+        }
+
+        public void setNext(Node<T> next) {
+            this.next = next;
+        }
+
+        public Node<T> getPrev() {
+            return prev;
+        }
+
+        public void setPrev(Node<T> prev) {
+            this.prev = prev;
+        }
```

Comment:
we don't need getters and setters for `Node` class, since it is nested, we can call its fields directly